packages/browser-wallet/src/popup/pages/SignCIS3Message/SignCIS3Message.tsx

```diff
@@ -0,0 +1,176 @@
+import React, { useCallback, useContext, useEffect, useState } from 'react';
+import { useLocation } from 'react-router-dom';
+import clsx from 'clsx';
+import { stringify } from 'json-bigint';
+import { useTranslation } from 'react-i18next';
+import { useAtomValue, useSetAtom } from 'jotai';
+import { Buffer } from 'buffer/';
+import {
+    AccountAddress,
+    AccountTransactionSignature,
+    buildBasicAccountSigner,
+    ConcordiumGRPCClient,
+    ContractAddress,
+    ContractName,
+    deserializeTypeValue,
+    EntrypointName,
+    getUpdateContractParameterSchema,
+    ModuleReference,
+    signMessage,
+} from '@concordium/web-sdk';
+import { fullscreenPromptContext } from '@popup/page-layouts/FullscreenPromptLayout';
+import { usePrivateKey } from '@popup/shared/utils/account-helpers';
+import { displayUrl } from '@popup/shared/utils/string-helpers';
+import { TextArea } from '@popup/shared/Form/TextArea';
+import ConnectedBox from '@popup/pages/Account/ConnectedBox';
+import Button from '@popup/shared/Button';
+import { addToastAtom } from '@popup/state';
+import ExternalRequestLayout from '@popup/page-layouts/ExternalRequestLayout';
+import { grpcClientAtom } from '@popup/store/settings';
+import { SignMessageObject } from '@concordium/browser-wallet-api-helpers';
+
+type Props = {
+    onSubmit(signature: AccountTransactionSignature): void;
+    onReject(): void;
+};
+
+interface Location {
+    state: {
+        payload: {
+            accountAddress: string;
+            message: SignMessageObject;
+            url: string;
+            cis3ContractDetails: Cis3ContractDetailsObject;
+        };
+    };
+}
+
+type Cis3ContractDetailsObject = {
+    contractAddress: ContractAddress.Type;
+    contractName: ContractName.Type;
+    entrypointName: EntrypointName.Type;
+};
+
+type DeserializedMessageObject = {
+    payload: bigint[] | [];
+};
+
+async function parseMessage(
+    message: SignMessageObject,
+    client: ConcordiumGRPCClient,
+    cis3ContractDetails: Cis3ContractDetailsObject
+) {
+    const deserializedMessage = deserializeTypeValue(
+        Buffer.from(message.data, 'hex'),
+        Buffer.from(message.schema, 'base64')
+    ) as DeserializedMessageObject;
+
+    const { contractAddress, contractName, entrypointName } = cis3ContractDetails;
+    const instanceInfo = await client.getInstanceInfo(contractAddress);
+
+    const schema = await client.getEmbeddedSchema(ModuleReference.fromHexString(instanceInfo.sourceModule.moduleRef));
```

Comment:
**Suggestion:** Ignore the embedded schema for now and assume the dapp always provide it (part of `message`), since this is what is done everywhere else in the wallet.